File: scripts/compose.js
import fs from "fs/promises"
import path from "path"
import inquirer from "inquirer"

const contentTypes = ["quote", "article", "idea", "project", "tool", "design", "book"]

function generateFrontmatter(type, answers) {
  const date = new Date().toISOString().split("T")[0]
  let frontmatter = `---
    title: "${answers.title}"
    date: "${date}"
    updated: "${date}"
  `

  switch (type) {
    case "quote":
      frontmatter += `author: "${answers.author}"
    quote: "${answers.quote}"
    categories: ${JSON.stringify(answers.categories)}
    tags: ${JSON.stringify(answers.tags)}
`
      break
    case "article":
      frontmatter += `author: "${answers.author}"
    description: "${answers.description}"
    categories: ${JSON.stringify(answers.categories)}
    tags: ${JSON.stringify(answers.tags)}
`
      break
    case "idea":
      frontmatter += `author: "${answers.author}"
    summary: "${answers.summary}"
    categories: ${JSON.stringify(answers.categories)}
    tags: ${JSON.stringify(answers.tags)}
`
      break
    case "project":
      frontmatter += `description: "${answers.description}"
    status: "${answers.status}"
    technologies: ${JSON.stringify(answers.technologies)}
    image: "${answers.image}"
    link: "${answers.link}"
`
      break
    case "tool":
      frontmatter += `description: "${answers.description}"
    category: "${answers.category}"
    image: "${answers.image}"
    link: "${answers.link}"
`
      break
    case "design":
      frontmatter += `description: "${answers.description}"
    category: "${answers.category}"
    image: "${answers.image}"
`
      break
    case "book":
      frontmatter += `author: "${answers.author}"
    description: "${answers.description}"
    genre: "${answers.genre}"
    rating: ${answers.rating}
    image: "${answers.image}"
`
      break
  }

  frontmatter += "---\n\n# " + answers.title + "\n\nAdd your content here.\n"
  return frontmatter
}

async function createMdxFile(type, answers) {
  const slug = answers.title
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
  const directory = path.join(process.cwd(), type + "s")
  const filePath = path.join(directory, `${slug}.mdx`)

  try {
    await fs.access(directory)
  } catch (error) {
    console.error(`Error: Directory '${type}s' does not exist.`)
    process.exit(1)
  }

  try {
    await fs.access(filePath)
    console.error(`Error: File '${slug}.mdx' already exists in the '${type}s' directory.`)
    process.exit(1)
  } catch (error) {
    // File doesn't exist, we can proceed
  }

  const content = generateFrontmatter(type, answers)

  try {
    await fs.writeFile(filePath, content)
    console.log(`Successfully created ${type} file: ${filePath}`)
  } catch (error) {
    console.error("Error creating file:", error)
  }
}

async function promptForDetails(type) {
  const questions = [
    {
      type: "input",
      name: "title",
      message: `Enter the title for the new ${type}:`,
      validate: (input) => input.trim() !== "" || "Title cannot be empty",
    },
  ]

  if (["quote", "article", "idea"].includes(type)) {
    questions.push(
      {
        type: "input",
        name: "author",
        message: "Enter the author name:",
        default: "Tyler Andor",
      },
      {
        type: "input",
        name: "categories",
        message: "Enter categories (comma-separated):",
        filter: (input) =>
          input
            .split(",")
            .map((item) => item.trim())
            .filter(Boolean),
      },
      {
        type: "input",
        name: "tags",
        message: "Enter tags (comma-separated):",
        filter: (input) =>
          input
            .split(",")
            .map((item) => item.trim())
            .filter(Boolean),
      },
    )
  }

  switch (type) {
    case "article":
    case "project":
    case "book":
      questions.push({
        type: "input",
        name: "description",
        message: "Enter a brief description:",
      })
      break
    case "idea":
      questions.push({
        type: "input",
        name: "summary",
        message: "Enter a brief summary:",
      })
      break
  }

  if (type === "quote") {
    questions.push({
      type: "input",
      name: "quote",
      message: "Enter the quote:",
    })
  }

  if (type === "project") {
    questions.push(
      {
        type: "input",
        name: "status",
        message: "Enter the project status:",
        default: "In Progress",
      },
      {
        type: "input",
        name: "technologies",
        message: "Enter technologies used (comma-separated):",
        filter: (input) =>
          input
            .split(",")
            .map((item) => item.trim())
            .filter(Boolean),
      },
    )
  }

  if (["project", "tool", "design", "book"].includes(type)) {
    questions.push({
      type: "input",
      name: "image",
      message: "Enter the image path:",
      default: "/placeholder.svg?height=600&width=1200",
    })
  }

  if (["project", "tool"].includes(type)) {
    questions.push({
      type: "input",
      name: "link",
      message: "Enter the project/tool link:",
    })
  }

  if (["tool", "design"].includes(type)) {
    questions.push(
      {
        type: "input",
        name: "description",
        message: "Enter a brief description:",
      },
      {
        type: "input",
        name: "category",
        message: "Enter a category:",
      },
    )
  }

  if (type === "book") {
    questions.push(
      {
        type: "input",
        name: "author",
        message: "Enter the author name:",
      },
      {
        type: "input",
        name: "genre",
        message: "Enter the book genre:",
      },
      {
        type: "number",
        name: "rating",
        message: "Enter the book rating (0-5):",
        validate: (input) => (input >= 0 && input <= 5) || "Rating must be between 0 and 5",
      },
    )
  }

  return inquirer.prompt(questions)
}

async function main() {
  try {
    const { type } = await inquirer.prompt([
      {
        type: "list",
        name: "type",
        message: "Select the content type:",
        choices: contentTypes,
      },
    ])

    const answers = await promptForDetails(type)
    await createMdxFile(type, answers)
  } catch (error) {
    console.error("An error occurred:", error)
  }
}

main()

// Example usage:
// node compose.js

File: scripts/compose.py
from textual.app import App, ComposeResult
from textual.widgets import Header, Footer, Label, Input, Button, OptionList
from textual.screen import Screen
from textual.containers import Vertical, Horizontal
import os
from pathlib import Path
import json
from datetime import datetime
import subprocess

CONTENT_TYPES = ["post", "quote", "article", "idea", "project", "tool", "design", "book"]
AUTHOR_NAME = "Tyler Andor"

def generate_frontmatter(content_type: str, answers: dict) -> str:
    """Generates the frontmatter content based on the content type and answers."""
    date = datetime.now().isoformat().split("T")[0]
    frontmatter = f"""---
title: "{answers.get('title', '')}"
author: "{AUTHOR_NAME}"
date: "{date}"
categories: {json.dumps([cat.strip() for cat in answers.get('categories', '').split(',') if cat.strip()])}
tags: {json.dumps([tag.strip() for tag in answers.get('tags', '').split(',') if tag.strip()])}
"""
    match content_type:
        case "post":
            frontmatter += f"description: \"{answers.get('description', '')}\"\n"
        case "quote":
            frontmatter += f"author: \"{answers.get('author', '')}\"\n"
        case "article":
            frontmatter += f"author: \"{answers.get('author', '')}\"\ndescription: \"{answers.get('description', '')}\"\n"
        case "idea":
            frontmatter += f"summary: \"{answers.get('summary', '')}\"\n"
        case "project":
            frontmatter += f"""description: "{answers.get('description', '')}"
status: "{answers.get('status', 'In Progress')}"
technologies: {json.dumps([tech.strip() for tech in answers.get('technologies', '').split(',') if tech.strip()])}
image: "{answers.get('image', '/placeholder.svg?height=600&width=1200')}"
link: "{answers.get('link', '')}"
"""
        case "tool":
            frontmatter += f"""description: "{answers.get('description', '')}"
image: "{answers.get('image', '/placeholder.svg?height=600&width=1200')}"
link: "{answers.get('link', '')}"
"""
        case "design":
            frontmatter += f"""description: "{answers.get('description', '')}"
image: "{answers.get('image', '/placeholder.svg?height=600&width=1200')}"
"""
        case "book":
            frontmatter += f"""author: "{answers.get('author', '')}"
description: "{answers.get('description', '')}"
genre: "{answers.get('genre', '')}"
rating: {answers.get('rating', 0)}
image: "{answers.get('image', '/placeholder.svg?height=600&width=1200')}"
"""
    frontmatter += "---\n\n# " + answers.get('title', '') + "\n\nAdd your content here.\n"
    return frontmatter

def create_mdx_file(content_type: str, answers: dict) -> Path | None:
    """Creates the MDX file with the generated frontmatter and returns the file path."""
    slug = answers.get('title', '').lower().replace(r'[^a-z0-9]+', '-').replace(r'^-+|-+$', '')
    directory = Path(os.getcwd()) / ".." / (content_type + "s")
    file_path = directory / f"{slug}.mdx"

    if not directory.exists():
        print(f"Error: Directory '{content_type}s' does not exist.")
        return None

    if file_path.exists():
        print(f"Error: File '{slug}.mdx' already exists in the '{content_type}s' directory.")
        return None

    content = generate_frontmatter(content_type, answers)

    try:
        with open(file_path, "w") as f:
            f.write(content)
        print(f"Successfully created {content_type} file: {file_path}")
        return file_path
    except Exception as e:
        print(f"Error creating file: {e}")
        return None

class ContentTypeScreen(Screen):
    """Screen for selecting the content type."""

    def compose(self) -> ComposeResult:
        yield Header()
        yield Label("Select the content type:")
        options = CONTENT_TYPES  # Just use the list of strings
        yield OptionList(*options, id="content-type-list")
        yield Footer()

    def on_option_list_selected(self, event: OptionList.OptionSelected) -> None:
        """Called when an option is selected (Enter key pressed)."""
        selected_type = str(event.option)  # Extract selected type as a string
        self.app.push_screen(DetailFormScreen(content_type=selected_type))

class DetailFormScreen(Screen):
    """Screen for entering details based on the content type."""

    def __init__(self, content_type: str, **kwargs) -> None:
        super().__init__(**kwargs)
        self.content_type = content_type
        self.inputs = {}

    def compose(self) -> ComposeResult:
        yield Header()
        yield Label(f"Enter details for: {self.content_type.capitalize()}")
        with Vertical(id="detail-inputs"):
            with Horizontal():
                yield Label("Title:")
                self.inputs["title"] = Input(id="title")
                yield self.inputs["title"]
            with Horizontal():
                yield Label("Categories (comma-separated):")
                self.inputs["categories"] = Input(id="categories")
                yield self.inputs["categories"]
            with Horizontal():
                yield Label("Tags (comma-separated):")
                self.inputs["tags"] = Input(id="tags")
                yield self.inputs["tags"]

            if self.content_type in ["post", "article", "project", "tool", "design", "book"]:
                with Horizontal():
                    yield Label("Description:")
                    self.inputs["description"] = Input(id="description")
                    yield self.inputs["description"]
            elif self.content_type == "idea":
                with Horizontal():
                    yield Label("Summary:")
                    self.inputs["summary"] = Input(id="summary")
                    yield self.inputs["summary"]

            if self.content_type in ["quote", "article", "book"]:
                with Horizontal():
                    yield Label("Author:")
                    self.inputs["author"] = Input(id="author")
                    yield self.inputs["author"]

            if self.content_type == "project":
                with Horizontal():
                    yield Label("Status:")
                    self.inputs["status"] = Input(id="status", value="In Progress")
                    yield self.inputs["status"]
                with Horizontal():
                    yield Label("Technologies (comma-separated):")
                    self.inputs["technologies"] = Input(id="technologies")
                    yield self.inputs["technologies"]
                with Horizontal():
                    yield Label("Image Path:")
                    self.inputs["image"] = Input(id="image", value="/placeholder.svg?height=600&width=1200")
                    yield self.inputs["image"]
                with Horizontal():
                    yield Label("Link:")
                    self.inputs["link"] = Input(id="link")
                    yield self.inputs["link"]

            elif self.content_type == "tool":
                with Horizontal():
                    yield Label("Image Path:")
                    self.inputs["image"] = Input(id="image", value="/placeholder.svg?height=600&width=1200")
                    yield self.inputs["image"]
                with Horizontal():
                    yield Label("Link:")
                    self.inputs["link"] = Input(id="link")
                    yield self.inputs["link"]

            elif self.content_type == "design":
                with Horizontal():
                    yield Label("Image Path:")
                    self.inputs["image"] = Input(id="image", value="/placeholder.svg?height=600&width=1200")
                    yield self.inputs["image"]

            elif self.content_type == "book":
                with Horizontal():
                    yield Label("Genre:")
                    self.inputs["genre"] = Input(id="genre")
                    yield self.inputs["genre"]
                with Horizontal():
                    yield Label("Rating (0-5):")
                    self.inputs["rating"] = Input(id="rating")
                    yield self.inputs["rating"]
                with Horizontal():
                    yield Label("Image Path:")
                    self.inputs["image"] = Input(id="image", value="/placeholder.svg?height=600&width=1200")
                    yield self.inputs["image"]

        yield Button("Create File", id="create-button", variant="primary")
        yield Footer()

    def on_button_pressed(self, event: Button.Pressed) -> None:
        """Event handler called when the create button is pressed."""
        if event.button.id == "create-button":
            answers = {}
            for input_id, input_widget in self.inputs.items():
                answers[input_id] = input_widget.value

            new_file_path = create_mdx_file(self.content_type, answers)
            if new_file_path:
                try:
                    subprocess.run(["nvim", str(new_file_path)])
                except FileNotFoundError:
                    print("Error: Neovim ('nvim') not found in your system's PATH.")
                except Exception as e:
                    print(f"Error opening file with Neovim: {e}")
            self.app.pop_screen()

class ContentCreatorApp(App):
    """A Textual app for creating MDX files with frontmatter."""

    SCREENS = {"content_type": ContentTypeScreen}
    BINDINGS = [
        ("q", "quit", "Quit"),
    ]

    def on_mount(self) -> None:
        self.push_screen("content_type")

if __name__ == "__main__":
    app = ContentCreatorApp()
    app.run()

